src/cli.ts

```diff
@@ -0,0 +1,181 @@
+import * as os from "os";
+
+import { validatePaths } from "vs/code/node/paths";
+import { parseMainProcessArgv } from "vs/platform/environment/node/argvHelper";
+import { ParsedArgs } from "vs/platform/environment/common/environment";
+import { buildHelpMessage, buildVersionMessage, options } from "vs/platform/environment/node/argv";
+import pkg from "vs/platform/product/node/package";
+import product from "vs/platform/product/node/product";
+
+import { MainServer } from "vs/server/src/server";
+import "vs/server/src/tar";
+import { AuthType, buildAllowedMessage, generateCertificate, generatePassword, open, unpackExecutables } from "vs/server/src/util";
+
+interface Args extends ParsedArgs {
+	auth?: AuthType;
+	"base-path"?: string;
+	cert?: string;
+	"cert-key"?: string;
+	"extra-builtin-extensions-dir"?: string;
+	"extra-extensions-dir"?: string;
+	host?: string;
+	open?: string;
+	port?: string;
+	socket?: string;
+}
+
+// The last item is _ which is like -- so our options need to come before it.
+const last = options.pop()!;
+
+// Remove options that won't work or don't make sense.
+let i = options.length;
+while (i--) {
+	switch (options[i].id) {
+		case "add":
+		case "diff":
+		case "file-uri":
+		case "folder-uri":
+		case "goto":
+		case "new-window":
+		case "reuse-window":
+		case "wait":
+		case "disable-gpu":
+		// TODO: pretty sure these don't work but not 100%.
+		case "max-memory":
+		case "prof-startup":
+		case "inspect-extensions":
+		case "inspect-brk-extensions":
+			options.splice(i, 1);
+			break;
+	}
```

Comment:
Should we add a `default` here that throws? Otherwise we'd loop forever if they add a new option.

Reply:
This shouldn't loop forever because of the `(i--)` right? If we `throw` on `default` then the app would crash on the first non-matching option (there will be a bunch since this is removing a subset of options that we don't or can't support).